Core/Util/GameManager.cpp

```diff
@@ -108,94 +109,207 @@ bool GameManager::Uninstall(std::string name) {
 				curDownload_.reset();
 				return;
 			}
 			// Game downloaded to temporary file - install it!
-			InstallGameOnThread(zipName, true);
+			InstallGameOnThread(curDownload_->url(), fileName, true);
 		} else {
-			ERROR_LOG(HLE, "Expected HTTP status code 200, got status code %i. Install cancelled, deleting partial file.", curDownload_->ResultCode());
-			File::Delete(zipName.c_str());
+			ERROR_LOG(HLE, "Expected HTTP status code 200, got status code %d. Install cancelled, deleting partial file '%s'",
+				curDownload_->ResultCode(), fileName.c_str());
+			File::Delete(fileName.c_str());
 		}
 		curDownload_.reset();
 	}
 }
 
-bool GameManager::InstallGame(std::string zipfile, bool deleteAfter) {
-	if (installInProgress_) {
-		ERROR_LOG(HLE, "Cannot have two installs in progress at the same time");
-		return false;
-	}
-
-	if (!File::Exists(zipfile)) {
-		ERROR_LOG(HLE, "ZIP file %s doesn't exist", zipfile.c_str());
-		return false;
+void countSlashes(std::string fileName, int *slashLocation, int *slashCount) {
+	*slashCount = 0;
+	int lastSlashLocation = -1;
+	*slashLocation = -1;
+	for (size_t i = 0; i < fileName.size(); i++) {
+		if (fileName[i] == '/') {
+			(*slashCount)++;
+			*slashLocation = lastSlashLocation;
+			lastSlashLocation = (int)i;
+		}
 	}
+}
 
-	I18NCategory *sy = GetI18NCategory("System");
-	installInProgress_ = true;
-
-	std::string pspGame = GetSysDirectory(DIRECTORY_GAME);
-	INFO_LOG(HLE, "Installing %s into %s", zipfile.c_str(), pspGame.c_str());
-	int error;
+ZipFileContents DetectZipFileContents(std::string fileName, ZipFileInfo *info) {
+	int error = 0;
 #ifdef _WIN32
-	struct zip *z = zip_open(ConvertUTF8ToWString(zipfile).c_str(), 0, &error);
+	struct zip *z = zip_open(ConvertUTF8ToWString(fileName).c_str(), 0, &error);
 #else
-	struct zip *z = zip_open(zipfile.c_str(), 0, &error);
+	struct zip *z = zip_open(fileName.c_str(), 0, &error);
 #endif
 	if (!z) {
-		ERROR_LOG(HLE, "Failed to open ZIP file %s, error code=%i", zipfile.c_str(), error);
-		return false;
+		return ZipFileContents::UNKNOWN;
 	}
+	ZipFileContents retVal = DetectZipFileContents(z, info);
+	zip_close(z);
+	return retVal;
+}
 
+ZipFileContents DetectZipFileContents(struct zip *z, ZipFileInfo *info) {
 	int numFiles = zip_get_num_files(z);
 
-	// First, find all the directories, and precreate them before we fill in with files.
-	// Also, verify that this is a PSP zip file with the correct layout.
-	bool isPSP = false;
+	// Verify that this is a PSP zip file with the correct layout. We also try
+	// to detect simple zipped ISO files, those we'll just "install" to the current
+	// directory of the Games tab (where else?).
+	bool isPSPMemstickGame = false;
+	bool isZippedISO = false;
 	int stripChars = 0;
+	int isoFileIndex = -1;
 
 	for (int i = 0; i < numFiles; i++) {
 		const char *fn = zip_get_name(z, i, 0);
 		std::string zippedName = fn;
 		if (zippedName.find("EBOOT.PBP") != std::string::npos) {
```

Comment:
I wonder if we ever have to think about `eboot.pbp` since the PSP wasn't case sensitive.

-[Unknown]

Reply:
Never seen one, but I'll make it handle lower-case.